design/ceph-managed-disruptionbudgets.md

```diff
@@ -0,0 +1,44 @@
+# Handling node drains through managed PodDisruptionBudgets.
+
+## Goals
+
+- Handle and block node drains that would cause data unavailablity and loss.
+- Unblock drains dynamically so that a rolling upgrade is made possible.
+- Allow for rolling upgrade of nodes in automated kubernetes environments like [cluster-api](https://github.com/kubernetes-sigs/cluster-api)
+
+
+
+## Design
+
+### OSDs
+
+OSDs do not fit under the single PodDisruptionBudget pattern. Ceph's ability to tolerate pod disruptions in one failure domain is dependent on the overall health of the cluster.
+Therefore, we will create a PodDisruptionBudget per failure domain that does not allow any evictions by default.
+When attempts to drain are detected, we will delete PodDisruption budget on one node at a time, progressing to the next only after ceph is healthy enough.
+The failure domain will be determined by the smallest failure domain of all the Ceph Pools in that cluster.
+
+Detecting drains is not easy as they are a client side operation. The client cordons the node and continuously attempts to evict all pods from the node until it succeeds.
+We will use a heuristic to detect drains. We will create a canary deployment for each node with a nodeSelector for that node. Since it is likely that pod will only be
+removed from the node in the event of a drain, we will rely on the assumption that if that pod is not running, that node is being drained. This will not be a dangerous assumption
+as false positives for drains are not dangerous in this use case.
+
+Example flow:
+- A Ceph pool CRD is created.
+- The controller creates a PDB with maxUnvailable of 0 for each failure domain.
+- A cluster upgrade agent wants to perform a kernel upgrade on the nodes.
+- It attempts to drain 1 or more nodes.
```

Comment:
There are a number of questions on the approach that I'm not seeing how it will be feasible to use the PDB for rook's purposes. A prototype may help answer most of these questions before any coding starts...
- If the maxUnavailable is 0, won't that prevent the canary pod from being evicted? If the drain does proceed to stop the canary pod, I would think the PDB was violated since it isn't supposed to allow any to be down.
- If the canary pod is stopped on the node where the rook operator is running, the operator will also be evicted and won't be able to respond to the event until the operator restarts elsewhere. This would be a good case for an HA operator, which has been discussed in the past, but not implemented yet.
- How will we tell the difference between a node being drained and if the user just wants to mark the node as unschedulable to new pods? Not sure that will be a common scenario, but I'm sure we will find other common scenarios where the heuristic thinks the node is being drained when it is not. If we have a false positive like this, what is the side effect? Will it just block real drains from proceeding?
- If all OSDs in a failure domain are down, new writes would be potentially blocked. For example, if there are three zones and the pool has replica 3, writes to each zone would be required for committing the write. So it isn't really as simple as limiting the evictions to a single failure domain. Thus, I go back to the simplest principle, which is to assume we're just taking down a single node at a time, then we can generalize the solution from there to multiple nodes in a failure domain for large clusters.
- Why do we need a canary pod? Can't the operator simply watch the existing osd pods? They're the pods we care about anyway.
- The admin might want to proceed with an upgrade even if the cluster is not healthy. For example, a newer version of K8 could actually fix an issue the customer is having. We need to allow the admin to force the upgrade.

Reply:
Had a discussion with @rohantmp to clarify these points. Overall, here are some clarifications to my questions:
1) The PDB is on the OSD pods, not the canary pod. When the canary pod is evicted, then the operator will determine if it should remove the PDB for the OSDs that would be affected on that node. Otherwise, the drain of that node would be stalled while the OSDs are not allowed to be evicted.
2) The operator can start elsewhere, there will just be a small delay before the operator starts up again and responds 
3) The heuristic will be imperfect, but false positives are ok to some degree
4) The operator could manage multiple PDBs for each zone. Thus, clusters with multiple nodes in each failure domain would be protected. If there is only a single node in a failure domain we wouldn't be protected, however, since writes would be blocked.
5) The PDB needs to be on a pod with a predictable count per node. In this case, the canary pod gives us a known single pod per node
6) The admin could force an upgrade despite the PDBs.

We should still have a prototype of this before coding starts in order to confirm behaviors around PDBs.

This is also a pattern that would be useful to other storage providers. This should be implemented in a way that can be re-used by others.

Reply:
If by prototype you mean to attempt a one-by-one drain while carrying out the steps mentioned manually to confirm that they have the expected behaviors, then I have carried this out already.

I have some scripts to label the pods by node and create per-node PDBs for anyone who wants to try this out: https://github.com/rohantmp/managed_pdb